### PR TITLE
Preempt old preparation

### PR DESCRIPTION
**Pull request by kdnilsen**
Previously, all old-gen regions not selected as collection set candidates were "made parseable" during the final-mark safepoint.   Making each of these regions parseable involves coalescing consecutive unmarked objects into contiguous larger objects that contain no pointers.  The process of identifying each span of consecutive unmarked objects within a large number of large heap regions was found to require hundreds of ms in some cases.  This patch changes the behavior so that preparation of heap regions for subsequent evacuation is performed during a concurrent phase of operation.  This concurrent phase must complete before any of the old heap regions placed in the candidate collection set can be evacuated.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Reviewers
 * [William Kemper](https://openjdk.java.net/census#wkemper) (@earthling-amzn - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/shenandoah pull/88/head:pull/88` \
`$ git checkout pull/88`

Update a local copy of the PR: \
`$ git checkout pull/88` \
`$ git pull https://git.openjdk.java.net/shenandoah pull/88/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 88`

View PR using the GUI difftool: \
`$ git pr show -t 88`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/shenandoah/pull/88.diff">https://git.openjdk.java.net/shenandoah/pull/88.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp [hunk 1]
```diff
@@ -72,6 +72,21 @@ class ShenandoahBreakpointMarkScope : public StackObj {
   }
 };
 
+class ShenandoahGlobalCoalesceAndFill : public ShenandoahHeapRegionClosure {
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:75]: Not sure how this ended up `shenandoahConcurrentGC.cpp`, last I checked it was defined in `shenandoahHeap.cpp`.

File: src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp [hunk 2]
```diff
@@ -171,6 +171,10 @@ HeapWord* ShenandoahFreeSet::try_allocate_in(ShenandoahHeapRegion* r, Shenandoah
 
     r->set_affiliation(req.affiliation());
     r->set_update_watermark(r->bottom());
+
+    // Any OLD region allocated during concurrent coalesce-and-fill does not need to be coalesced and filled.
+    // This code is only necessary if req.affiliation() is old, but harmless if not.
+    r->finish_coalesce_and_fill();
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp]: Not sure I understand why that's the case. This region could still contain garbage objects that could break the remembered set scan. How does allocating out of the region make these objects parseable?

File: src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp [hunk 3]
```diff
@@ -458,7 +464,17 @@ void ShenandoahHeapRegion::oop_fill_and_coalesce() {
       heap->card_scan()->coalesce_objects(obj_addr, fill_size);
       obj_addr = next_marked_obj;
     }
+    if (!ops_before_preempt_check--) {
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp:467]: I think the `C++` way is to compare against 0, not coerce to bool.

**Review by earthling-amzn — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp (hunk 1 above)
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:75]: Not sure either.  It is in both places in my branch.  I see no reason for it to be here so am removing it.

File: src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp (hunk 3 above)
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp:467]: Thanks.  I've changed the condition.

**Pull request closed without merging**